apps/system/js/sync_manager.js

```diff
@@ -0,0 +1,471 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict';
+
+/* global BaseModule */
+/* global ERROR_GET_FXA_ASSERTION */
+/* global ERROR_REQUEST_SYNC_REGISTRATION */
+/* global ERROR_SYNC_APP_GENERIC */
+/* global ERROR_SYNC_APP_KILLED */
+/* global ERROR_SYNC_REQUEST */
+/* global FxAccountsClient */
+/* global LazyLoader */
+/* global Service */
+/* global SettingsHelper */
+/* global SyncUnrecoverableErrors */
+/* global uuid */
+
+/**
+ * This module manages the lifecycle of Firefox Sync on Firefox OS.
+ * It connects the Settings application where the Firefox Sync UI is presented
+ * to the user with the Sync app that is responsible of doing the actual
+ * synchronization of the user's data.
+ *
+ * SyncManager reacts to changes done on the Settings Sync UI that triggers
+ * IAC requests or Settings API changes and translates these changes into
+ * actions. It makes use of the SyncStateMachine to control that this actions
+ * are possible on the current situation of Sync and communicates with the
+ * Sync application via IAC API messages when needed.
+ */
+
+(function() {
+  const FXA_EVENT = 'mozFxAccountsUnsolChromeEvent';
+
+  const SYNC_TASK = 'gaia::system::firefoxsync';
+  const SYNC_REQUEST_IAC_KEYWORD = 'gaia::sync::request';
+
+  const HISTORY = 'history';
+  const PASSWORDS = 'passwords';
+
+  var SyncManager = function() {};
+
+  SyncManager.STATES = [];
+
+  SyncManager.SERVICES = [];
+
+  SyncManager.SETTINGS = [
+    /** READ ONLY SETTINGS **/
+    // These settings must only be changed by the Settings app and read
+    // by the System app.
+
+    // The user can choose the collections she wants to sync. These settings
+    // indicates the user choice per collection.
+    'sync.collection.history.enabled',
+    'sync.collection.passwords.enabled',
+
+
+    /** READ WRITE SETTINGS **/
+    // Can be 'disabled', 'enabling', 'enabled', 'syncing' or 'errored'.
+    // No need to observe, we will only be reading it once on _start.
+    // -- 'sync.state',
+
+    /** WRITE ONLY SETTINGS **/
+    // ( No need to be observed )
+    // These settings must only be changed by the System app and read
+    // by the Settings app.
+
+    // Indicates the last error in case that 'sync.state' is 'errored'.
+    // -- 'sync.state.error',
+
+    // Last time we synchronized successfully.
+    // -- 'sync.lastTime'
+  ];
```

Comment:
The three properties ('sync.state', 'sync.state.error', 'sync.lastTime') look like read-only properties for Settings app (or other apps which would like to control fxsync), but these properties could be changed by any apps which have mozSettings permission. The state will probably be ruined by other apps.
Can these state/lastTime information be retrieved via IAC way just like iac-gaia::sync::management?